### ISSUE TITLE
error in tests building for riscv

### ISSUE THREAD

**Issue by itsshowtime** [xianyi/OpenBLAS#3294]
I am trying to build **OpenBLAS 0.3.15** in a risc-v board called BeagleV, the processor is a RISC-V U74 Dual core.

The command line I used is the following:
```
$ make CC=gcc FC=gfortran BINARY=64 COMMON_OPT="-O3 -fopenmp" FCOMMON_OPT="-O3 -fopenmp -frecursive" USE_OPENMP=1 TARGET=RISCV64_GENERIC PREFIX=... all
```

I get the following error when the tests are running:
```
make -j 2 -C utest all
make[1]: Entering directory '/home/kpeiro/ss/OpenBLAS-0.3.15/utest'
./openblas_utest
TEST 1/35 max:smax_zero [OK]
TEST 2/35 max:dmax_positive [OK]
TEST 3/35 max:smax_negative [OK]
TEST 4/35 min:smin_zero [OK]
TEST 5/35 min:dmin_positive [OK]
TEST 6/35 min:smin_negative [OK]
TEST 7/35 amax:damax [OK]
TEST 8/35 amax:samax [OK]
TEST 9/35 ismax:negative_step_2 [OK]
TEST 10/35 ismax:positive_step_2 [OK]
TEST 11/35 ismin:negative_step_2 [FAIL]
  ERR: test_ismin.c:63  expected 9, got 1
TEST 12/35 ismin:positive_step_2 [FAIL]
  ERR: test_ismin.c:50  expected 9, got 50
TEST 13/35 drotmg:drotmg_D1_big_D2_big_flag_zero [OK]
TEST 14/35 drotmg:rotmg_D1eqD2_X1eqX2 [OK]
TEST 15/35 drotmg:rotmg_issue1452 [OK]
TEST 16/35 drotmg:rotmg [OK]
TEST 17/35 axpy:caxpy_inc_0 [OK]
TEST 18/35 axpy:saxpy_inc_0 [OK]
TEST 19/35 axpy:zaxpy_inc_0 [OK]
TEST 20/35 axpy:daxpy_inc_0 [OK]
TEST 21/35 zdotu:zdotu_offset_1 [OK]
TEST 22/35 zdotu:zdotu_n_1 [OK]
TEST 23/35 dsdot:dsdot_n_1 [FAIL]
  ERR: test_dsdot.c:47  expected -2.393e-03, got -2.393e-03 (diff -1.035e-10, tol 1.000e-13)
TEST 24/35 swap:cswap_inc_0 [OK]
TEST 25/35 swap:sswap_inc_0 [OK]
TEST 26/35 swap:zswap_inc_0 [OK]
TEST 27/35 swap:dswap_inc_0 [OK]
TEST 28/35 rot:csrot_inc_0 [OK]
TEST 29/35 rot:srot_inc_0 [OK]
TEST 30/35 rot:zdrot_inc_0 [OK]
TEST 31/35 rot:drot_inc_0 [OK]
TEST 32/35 potrf:smoketest_trivial [OK]
TEST 33/35 potrf:bug_695 [OK]
TEST 34/35 kernel_regress:skx_avx [OK]
TEST 35/35 fork:safety_after_fork_in_parent [OK]
RESULTS: 35 tests (32 ok, 3 failed, 0 skipped) ran in 291 ms
make[1]: *** [Makefile:52: run_test] Error 3
make[1]: Leaving directory '/home/kpeiro/ss/OpenBLAS-0.3.15/utest'
make: *** [Makefile:147: tests] Error 2
```

Since I get these errors, the file `libopenblas_riscv64_genericp-r0.3.15.so` is not generated.
I have also tried to compile with a SiFive U54 processor, getting the same error.

Am I doing something wrong?  Could I please get any help?

**Comment by brada4** [comment]
What happens without any COMMON_OPT ? Just USE_OPENMP=1 instead? -O3 introduces non-conformant numeric optimisations.

**Comment by martin-frbg** [comment]
Yes please try with `-O2` instead of the `-O3` - the generic RISCV target uses plain C kernels and DSDOT is just s a fairly obvious unrolled loop doing sum+=(double)a*(double)b, so this loss of precision is probably down to the compiler cutting corners somewhere.

**Comment by brada4** [comment]
I think it also overrides -fno-optimize-sibling-calls for fortran potentially yielding spectacular crashes due to ABI mismatch when actual fortran routines get called.

**Comment by itsshowtime** [comment]
Okay I'm building now with -O2 so I will have more news later or tomorrow. Thank you all.

**Comment by brada4** [comment]
You do not need to override the flags as long as default flags work for you.

**Comment by itsshowtime** [comment]
I got the same results with `-O2`

```
make CC=gcc FC=gfortran BINARY=64 COMMON_OPT="-O2 -fopenmp" FCOMMON_OPT="-O2 -fopenmp -frecursive" USE_OPENMP=1 TARGET=RISCV64_GENERIC PREFIX=/apps/riscv/openBLAS/0.3.15 all
```
```
./openblas_utest
TEST 1/35 max:smax_zero [OK]
TEST 2/35 max:dmax_positive [OK]
TEST 3/35 max:smax_negative [OK]
TEST 4/35 min:smin_zero [OK]
TEST 5/35 min:dmin_positive [OK]
TEST 6/35 min:smin_negative [OK]
TEST 7/35 amax:damax [OK]
TEST 8/35 amax:samax [OK]
TEST 9/35 ismax:negative_step_2 [OK]
TEST 10/35 ismax:positive_step_2 [OK]
TEST 11/35 ismin:negative_step_2 [FAIL]
  ERR: test_ismin.c:63  expected 9, got 1
TEST 12/35 ismin:positive_step_2 [FAIL]
  ERR: test_ismin.c:50  expected 9, got 50
TEST 13/35 drotmg:drotmg_D1_big_D2_big_flag_zero [OK]
TEST 14/35 drotmg:rotmg_D1eqD2_X1eqX2 [OK]
TEST 15/35 drotmg:rotmg_issue1452 [OK]
TEST 16/35 drotmg:rotmg [OK]
TEST 17/35 axpy:caxpy_inc_0 [OK]
TEST 18/35 axpy:saxpy_inc_0 [OK]
TEST 19/35 axpy:zaxpy_inc_0 [OK]
TEST 20/35 axpy:daxpy_inc_0 [OK]
TEST 21/35 zdotu:zdotu_offset_1 [OK]
TEST 22/35 zdotu:zdotu_n_1 [OK]
TEST 23/35 dsdot:dsdot_n_1 [FAIL]
  ERR: test_dsdot.c:47  expected -2.393e-03, got -2.393e-03 (diff -1.035e-10, tol 1.000e-13)
TEST 24/35 swap:cswap_inc_0 [OK]
TEST 25/35 swap:sswap_inc_0 [OK]
TEST 26/35 swap:zswap_inc_0 [OK]
TEST 27/35 swap:dswap_inc_0 [OK]
TEST 28/35 rot:csrot_inc_0 [OK]
TEST 29/35 rot:srot_inc_0 [OK]
TEST 30/35 rot:zdrot_inc_0 [OK]
TEST 31/35 rot:drot_inc_0 [OK]
TEST 32/35 potrf:smoketest_trivial [OK]
TEST 33/35 potrf:bug_695 [OK]
TEST 34/35 kernel_regress:skx_avx [OK]
TEST 35/35 fork:safety_after_fork_in_parent [OK]
RESULTS: 35 tests (32 ok, 3 failed, 0 skipped) ran in 277 ms
make[1]: *** [Makefile:52: run_test] Error 3
make[1]: Leaving directory '/home/kpeiro/ss/OpenBLAS-0.3.15/utest'
make: *** [Makefile:147: tests] Error 2
```

**Comment by brada4** [comment]
Could you please repeat tests with your overriden CFLAGS completely removed? Like typing `make clean; make` alone?

**Comment by itsshowtime** [comment]
> Could you please repeat tests with your overriden CFLAGS completely removed? Like typing `make clean; make` alone?

If I just do `make clean; make`, it does not compile the right target.

I have also tried by removing completely the CFLAGS and FFLAGS, with the same error:
```
make CC=gcc FC=gfortran BINARY=64 USE_OPENMP=1 TARGET=RISCV64_GENERIC PREFIX=/apps/riscv/openBLAS/0.3.15 all
```

I am trying now with just:
```
make USE_OPENMP=1 TARGET=RISCV64_GENERIC PREFIX=/apps/riscv/openBLAS/0.3.15 all
```

Thank you for all your suggestions! I will update you when I have more news.

**Comment by itsshowtime** [comment]
I confirm that compiling with:
```
make USE_OPENMP=1 TARGET=RISCV64_GENERIC PREFIX=/apps/riscv/openBLAS/0.3.15 all
```
gives me the same exact error.

**Comment by martin-frbg** [comment]
Very strange - there must be something seriously amiss with the platform, as I see no way that the C code could be wrong (ISMIN a.k.a kernel/risc64/imin.c is even more trivial than DSDOT). Does the openblas_utest program in the utest subdirectory run without
error when you invoke it with `OPENMP_NUM_THREADS=1 ./openblas_utest` ? (That would suggest some kind of thread race)

**Comment by itsshowtime** [comment]
Running directly with `OPENMP_NUM_THREADS=1 ./openblas_utest` gives me the same errors

**Comment by martin-frbg** [comment]
That would seem to rule out the one source of coding errors that I can think of. Any chance that your Beagleboard may be running too hot, or on an insufficient power supply ?

**Comment by itsshowtime** [comment]
Hmm I don't think so, and I have also tried it in another risc-v board (PolarFire) with the exact same results.

**Comment by martin-frbg** [comment]
I have no experience with RISC-V hardware myself but I struggle to see what could go wrong with a trivial C function in a single thread, unless either the compiler or the hardware is broken. The code used in RISCV64_GENERIC is shared with all the generic
targets of other architectures (and a few non-generic ones as well), where it has not caused any embarassment.

**Comment by martin-frbg** [comment]
Any warnings encountered during compilation ?

**Comment by itsshowtime** [comment]
No, I have attached the whole compilation log: [comp_output.txt](https://github.com/xianyi/OpenBLAS/files/6787207/comp_output.txt)

**Comment by brada4** [comment]
Hey, you filtered out warnings printed on stderr.
`make 2>&1 > ../compile_out.txt`
or
`script ../compile_out.txt`
would be able to capture them.`

**Comment by martin-frbg** [comment]
Which gcc version is this ? (Any chance to try a different/newer one ?) Obviously you could just hack the main Makefile and remove the utest (which mainly tests corner cases), but then I'd run at least a `make lapack-test` with all the other `make` options after the build. (This requires python for printing a summary of the results though)

**Comment by itsshowtime** [comment]
I am logging everything now, I will have results in some hours

The gcc version is **10.3.1**, I could try with **11.1** if needed.

**Comment by martin-frbg** [comment]
No idea what's needed, this simply is not expected to happen :frowning_face:  - other things to try would be compiling everything at `-O0`  or alternatively adding `#pragma GCC optimize "O0"` at the top of the affected files (kernel/generic/dot.c and kernel/riscv64/imin.c)

**Comment by itsshowtime** [comment]
There are warnings when compiling with `-O2` (whole compilation log attached)
[comp_output.txt](https://github.com/xianyi/OpenBLAS/files/6791348/comp_output.txt)

I am now trying with `-O0`.

**Comment by martin-frbg** [comment]
Nothing suspicious in that log (LAPACK build is notoriously noisy, and the OpenBLAS part has accumulated a few unused variables lately)

**Comment by itsshowtime** [comment]
Compiling with `-O0` gives me the same error.

I will now try with **gcc 11.1**, and if the error keeps appearing I will remove the `utest`.

**Comment by brada4** [comment]
The whole test is one single multiplication within BLAS library call.
particular value printed is:
```
     ((double)0.172555164F * (double)-0.0138700781F)-(-0.00239335360107)
```
It is expected to lose 1.5 bits of precision, while it seems that almost 10 bits got lost on RISCV

**Comment by martin-frbg** [comment]
Well, the ISMIN tests aren't expected to fail either, when using the trivial C implementation. The utests were put there to find incorrect optimizations in the hand-crafted assembly kernels. Something must be very wrong with that RISCV setup and I do not think it is OpenBLAS - either GCC or binutils/assembler.

**Comment by brada4** [comment]
GCC manual (info gcc) has things probably related in 
> 3.18.37 RISC-V Options
'-mfdiv'
'-mno-fdiv'
'-mdiv'
'-mno-div'

@itsshowtime  could you try with `COMMON_OPT="-O2 -mno-fdiv -mno-div` (and more -mno-whatever you find in local documentation)

EDIT: also try `-mno-save-restore` since library calls are involved, `-O2` can stay.

**Comment by brada4** [comment]
I was checking idea of rounding flag getting wrong, what it seems is a `(double)(a*b)` where `(double)(a)*(double)(b)` should have been, not moved by rounding modes at all.
`3294_rounding.c`
```
#include <stdio.h>
#include <fenv.h>
int main ()
{
        printf("%e\n",((double)0.172555164F * (double)-0.0138700781F)-(-0.00239335360107));
        printf("%e\n",(double)(0.172555164F * -0.0138700781F)-(-0.00239335360107));
        FE_TONEAREST;
        printf("%e\n",((double)0.172555164F * (double)-0.0138700781F)-(-0.00239335360107));
        printf("%e\n",(double)(0.172555164F * -0.0138700781F)-(-0.00239335360107));
        FE_UPWARD;
        printf("%e\n",((double)0.172555164F * (double)-0.0138700781F)-(-0.00239335360107));
        printf("%e\n",(double)(0.172555164F * -0.0138700781F)-(-0.00239335360107));
        FE_DOWNWARD;
        printf("%e\n",((double)0.172555164F * (double)-0.0138700781F)-(-0.00239335360107));
        printf("%e\n",(double)(0.172555164F * -0.0138700781F)-(-0.00239335360107));
        FE_TOWARDZERO;
        printf("%e\n",((double)0.172555164F * (double)-0.0138700781F)-(-0.00239335360107));
        printf("%e\n",(double)(0.172555164F * -0.0138700781F)-(-0.00239335360107));
}
```
[3294_rounding.txt](https://github.com/xianyi/OpenBLAS/files/6797189/3294_rounding.txt)

**Comment by brada4** [comment]
x86_64 , tried piledriver, zen1 , skylakex, all same.
```
-1.498801e-15
1.034605e-10
-1.498801e-15
1.034605e-10
-1.498801e-15
1.034605e-10
-1.498801e-15
1.034605e-10
-1.498801e-15
1.034605e-10
```

**Comment by martin-frbg** [comment]
Oh no, sorry - I was comparing current develop branch, where exactly these problems have already been fixed by #3259 - only a month ago, but I have been very busy lately and did not give riscv much attention until two weeks ago. Sorry for bothering you with pointless exercises, current develop branch or upcoming 0.3.16 should build just fine as RISCV64_GENERIC

**Comment by itsshowtime** [comment]
Hi all, I have successfully compiled OpenBLAS now, thank you!
